iceoryx_posh/test/moduletests/test_popo_toml_gateway_config_parser.cpp

```diff
@@ -20,201 +20,528 @@
-    }
+    auto result = StubbedTomlGatewayConfigParser::validate(*toml);
+    ASSERT_TRUE(result.has_error());
+    EXPECT_EQ(TomlGatewayConfigParseError::INCOMPLETE_SERVICE_DESCRIPTION, result.get_error());
 }
 
-TEST_F(TomlGatewayConfigParserTest, FailsValidationIfServiceDescriptionBeginsWithNumber)
+TEST_F(TomlGatewayConfigParserTest, ServiceDescriptionBeginsWithNumberReturnInvalidServiceDescriptionError)
 {
-    // ===== Setup
-    // Prepare configuration to test with
     auto toml = cpptoml::make_table();
     auto serviceArray = cpptoml::make_table_array();
 
     auto serviceBeginsWithNumber = cpptoml::make_table();
     serviceBeginsWithNumber->insert("service", "0000");
     serviceBeginsWithNumber->insert("instance", "0000");
     serviceBeginsWithNumber->insert("event", "0000");
-    serviceBeginsWithNumber->insert("size", 0);
     serviceArray->push_back(serviceBeginsWithNumber);
     toml->insert("services", serviceArray);
 
-    // ===== Test
+    auto result = StubbedTomlGatewayConfigParser::validate(*toml);
+    ASSERT_TRUE(result.has_error());
+    EXPECT_EQ(TomlGatewayConfigParseError::INVALID_SERVICE_DESCRIPTION, result.get_error());
+}
+
+TEST_F(TomlGatewayConfigParserTest, HyphenInServiceDescriptionReturnInvalidServiceDescriptionError)
+{
+    auto toml = cpptoml::make_table();
+    auto serviceArray = cpptoml::make_table_array();
+
+    auto serviceEntryWithHyphen = cpptoml::make_table();
+    serviceEntryWithHyphen->insert("service", "service-name");
+    serviceEntryWithHyphen->insert("instance", "instance-name");
+    serviceEntryWithHyphen->insert("event", "event-name");
+    serviceArray->push_back(serviceEntryWithHyphen);
+    toml->insert("services", serviceArray);
+
+    auto result = StubbedTomlGatewayConfigParser::validate(*toml);
+    ASSERT_TRUE(result.has_error());
+    EXPECT_EQ(TomlGatewayConfigParseError::INVALID_SERVICE_DESCRIPTION, result.get_error());
+}
+
+TEST_F(TomlGatewayConfigParserTest, NoServicesInConfigReturnIncompleteConfigurationError)
+{
+    auto toml = cpptoml::make_table();
+
     auto result = iox::config::StubbedTomlGatewayConfigParser::validate(*toml);
-    EXPECT_EQ(true, result.has_error());
-    if (result.has_error())
-    {
-        EXPECT_EQ(iox::config::TomlGatewayConfigParseError::INVALID_SERVICE_DESCRIPTION, result.get_error());
-    }
+
+    ASSERT_TRUE(result.has_error());
+    EXPECT_EQ(TomlGatewayConfigParseError::INCOMPLETE_CONFIGURATION, result.get_error());
+}
+
+TEST_F(TomlGatewayConfigParserTest, ParseWithoutParameterTakeDefaultPathReturnNoError)
+{
+    auto result = TomlGatewayConfigParser::parse();
+    GatewayConfig config = result.value();
+
+    EXPECT_FALSE(result.has_error());
+    EXPECT_TRUE(config.m_configuredServices.empty());
+}
+
+TEST_F(TomlGatewayConfigParserTest, ParseWithEmptyPathReturnEmptyConfig)
+{
+    iox::roudi::ConfigFilePathString_t path = "";
```

Comment:
This path is not empty because it creates a string with a null-terminator in it.

Reply:
Yeah, what i would to test here is no provided path, i will refactor the test name
![image](https://user-images.githubusercontent.com/77273698/109274620-01cfb280-7814-11eb-96b6-102ca97e69bb.png)

Reply:
The Problem is more that this test maybe goes into the right branch here so it is covered but it does no really verifies the behavior because the only indication for the empty path is a printed log message.

Reply:
@dkroenke, I understand , do you have any proposal ?

Reply:
@dkroenke i propose here when the path size is equal to zero or we have an invalid path , an error handler should be returned, to be able to test this two cases, what do you think?

Reply:
According to the guidelines in https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/design/error-handling.md the errorhandler should only be used in RouDi or the Runtime. The TomlGatewayParser is used in gateway executables so this is not an option.
An idea which came in my mind is that maybe we return a config in this case which indicates that something went wrong.
Basically there are two cases: 1. File path empty 2. config file could not be opened
If for example case 1 occur, then could maybe return a config with a service entry which has a ServiceDescription like this:
```
ServiceDescription("wrong_config_file_path", "wrong_config_file_path", "wrong_config_file_path")
```
Then we have an indication at loglevel and the returned config can be tested and tells the user that something went wrong.
@ithier Would it make sense to have that?

Reply:
What's wrong with just returning `expected::error`, like it already does ?

Reply:
If we want to inform the user of this, maybe we should change the behaviour to return `expected::error` in this case?

Then the caller - likely the main function of a gateway executable - could either terminate the program, or choose to proceed without a configuration.

Reply:
yeah that was my suggestion

Reply:
@ithier @manel94 i added this as open point to #574